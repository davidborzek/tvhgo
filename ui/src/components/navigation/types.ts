import { ReactElement } from 'react';

export type INavigationItem = {
  to: string;
  title: string;
  requiredRoles?: string[];
  icon?: ReactElement;
  items?: INavigationItem[];
};
