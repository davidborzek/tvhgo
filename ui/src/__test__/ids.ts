export enum TestIds {
  PAGINATION_FIRST_PAGE = 'pagination-first-page-btn',
  PAGINATION_PREVIOUS_PAGE = 'pagination-previous-page-btn',
  PAGINATION_NEXT_PAGE = 'pagination-next-page-btn',
  PAGINATION_LAST_PAGE = 'pagination-last-page-btn',
  REVOKE_SESSION_BUTTON = 'revoke-session-btn',
  REVOKE_TOKEN_BUTTON = 'revoke-token-btn',
  THEME_DROPDOWN = 'theme-dropdown',
  LANGUAGE_DROPDOWN = 'language-dropdown',
  LOGOUT_BUTTON = 'logout-button',
  SAVE_USER_BUTTON = 'save-user-btn',
}
