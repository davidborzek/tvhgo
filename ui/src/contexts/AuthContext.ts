import { createContext, useContext } from 'react';

import { UserResponse } from '@/clients/api/api.types';

export type AuthContextProps = {
  user: UserResponse | null;
  setUser: (user: UserResponse | null) => void;
};

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  setUser: () => {
    throw new Error('not implemented');
  },
});

export const useAuth = () => {
  return useContext(AuthContext);
};
