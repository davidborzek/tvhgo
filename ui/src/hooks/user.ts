import {
  ApiError,
  createUser,
  deleteUser,
  updateUser,
  updateUserPassword,
} from '@/clients/api/api';
import {
  CreateUser,
  UpdateUser,
  UpdateUserPassword,
} from '@/clients/api/api.types';

import { useAuth } from '@/contexts/AuthContext';
import { useNotification } from './notification';
import { useRevalidator } from 'react-router-dom';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';

export const useUpdateUser = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('manageUser');

  const { setUser } = useAuth();
  const { t } = useTranslation();

  const update = async (opts: UpdateUser) => {
    dismissNotification();

    await updateUser(opts)
      .then((user) => {
        notifySuccess(t('user_updated_successfully'));
        setUser(user);
      })
      .catch((error) => {
        if (error instanceof ApiError && error.code === 400) {
          if (error.message === 'username already exists') {
            notifyError(t('username_already_exists'));
          } else if (error.message === 'email already exists') {
            notifyError(t('email_already_exists'));
          }
          return;
        }

        notifyError(t('unexpected'));
      });
  };

  return { update };
};

export const useUpdateUserPassword = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('changeUserPassword');

  const { t } = useTranslation();

  const updatePassword = async (opts: UpdateUserPassword) => {
    dismissNotification();

    return await updateUserPassword(opts)
      .then(() => {
        notifySuccess(t('password_updated_successfully'));
        return true;
      })
      .catch((error) => {
        if (
          error instanceof ApiError &&
          error.code === 400 &&
          error.message === 'current password is invalid'
        ) {
          notifyError(t('invalid_current_password'));
          return false;
        }

        notifyError(t('unexpected'));
      });
  };

  return { updatePassword };
};

export const useCreateUser = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('createUser');

  const { t } = useTranslation();

  const create = async (opts: CreateUser) => {
    dismissNotification();

    return await createUser(opts)
      .then(() => {
        notifySuccess(t('user_created_successfully'));
      })
      .catch((error) => {
        if (error instanceof ApiError && error.code === 400) {
          if (error.message === 'username already exists') {
            notifyError(t('username_already_exists'));
          } else if (error.message === 'email already exists') {
            notifyError(t('email_already_exists'));
          }
        } else {
          notifyError(t('unexpected'));
        }

        throw error;
      });
  };

  return { create };
};

export const useDeleteUser = (): [boolean, (id: number) => Promise<void>] => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('deleteUser');
  const [isPending, setIsPending] = useState(false);
  const { revalidate } = useRevalidator();

  const { t } = useTranslation();

  const _deleteUser = async (id: number) => {
    dismissNotification();
    setIsPending(true);
    return deleteUser(id)
      .then(() => {
        notifySuccess(t('user_deleted_successfully'));
        revalidate();
      })
      .catch((error) => {
        if (error instanceof ApiError && error.code === 400) {
          if (error.message === 'current user cannot be deleted') {
            notifyError(t('current_user_cannot_be_deleted'));
          }
        } else {
          notifyError(t('unexpected'));
        }

        throw error;
      })
      .finally(() => setIsPending(false));
  };

  return [isPending, _deleteUser];
};
