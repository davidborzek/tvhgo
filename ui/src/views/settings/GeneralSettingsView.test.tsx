import { Theme, useTheme } from '@/contexts/ThemeContext';
import { afterEach, beforeEach, expect, test, vi } from 'vitest';
import { cleanup, render } from '@testing-library/react';
import { useLoaderData, useNavigate } from 'react-router-dom';

import { Component as GeneralSettingsView } from './GeneralSettingsView';
import { TestIds } from '@/__test__/ids';
import { UserResponse } from '@/clients/api/api.types';
import i18n from 'i18next';
import { useAuth } from '@/contexts/AuthContext';
import { useUpdateUser } from '@/hooks/user';
import { userEvent } from '@testing-library/user-event';

vi.mock('@/contexts/AuthContext');
vi.mock('@/contexts/ThemeContext');
vi.mock('@/hooks/user');
vi.mock('react-router-dom');
vi.mock('i18next');

const user: UserResponse = {
  createdAt: 0,
  displayName: 'Some User',
  email: 'some@email.com',
  id: 1,
  updatedAt: 0,
  username: 'someUsername',
};

afterEach(() => {
  vi.restoreAllMocks();
  cleanup();
});

const setThemeMock = vi.fn();
const navigateMock = vi.fn();
const updateUserMock = vi.fn();

beforeEach(() => {
  vi.mocked(useAuth).mockReturnValue({
    setUser: vi.fn(),
    user,
  });

  vi.mocked(useTheme).mockReturnValue({
    setTheme: setThemeMock,
    theme: Theme.DARK,
  });

  vi.mocked(useUpdateUser).mockReturnValue({
    update: updateUserMock,
  });

  vi.mocked(useNavigate).mockReturnValue(navigateMock);

  vi.mocked(useLoaderData).mockReturnValue([
    {
      forwardAuth: false,
      sessionId: 2,
      userId: 1,
    },
  ]);
});

test('should render', () => {
  const document = render(<GeneralSettingsView />);
  expect(document.asFragment()).toMatchSnapshot();
});

test('should render render when no user is present', () => {
  vi.mocked(useAuth).mockReturnValue({
    setUser: vi.fn(),
    user: null,
  });

  const document = render(<GeneralSettingsView />);
  expect(document.asFragment()).toMatchSnapshot();
});

test('should render render without logout button when authenticated by reverse proxy', () => {
  vi.mocked(useLoaderData).mockReturnValue([
    {
      forwardAuth: true,
      sessionId: null,
      userId: 1,
    },
  ]);

  const document = render(<GeneralSettingsView />);
  expect(document.asFragment()).toMatchSnapshot();
});

test('should update user', async () => {
  const document = render(<GeneralSettingsView />);

  const newUsername = 'newUsername';
  const newEmail = 'new@email.com';
  const newDisplayName = 'New Name';

  const usernameInput = document.container.querySelector(
    'input[name=username]'
  ) as Element;
  const emailInput = document.container.querySelector(
    'input[name=email]'
  ) as Element;
  const displayNameInput = document.container.querySelector(
    'input[name=displayName]'
  ) as Element;
  const saveButton = document.getByTestId(TestIds.SAVE_USER_BUTTON);

  await userEvent.clear(usernameInput);
  await userEvent.type(usernameInput, newUsername);
  await userEvent.clear(emailInput);
  await userEvent.type(emailInput, newEmail);
  await userEvent.clear(displayNameInput);
  await userEvent.type(displayNameInput, newDisplayName);

  expect(usernameInput).toHaveValue(newUsername);
  expect(emailInput).toHaveValue(newEmail);
  expect(displayNameInput).toHaveValue(newDisplayName);

  await userEvent.click(saveButton);

  expect(updateUserMock).toHaveBeenCalledWith({
    displayName: newDisplayName,
    email: newEmail,
    username: newUsername,
  });
});

test('should logout', async () => {
  const document = render(<GeneralSettingsView />);

  const logoutButton = document.getByTestId(TestIds.LOGOUT_BUTTON);
  await userEvent.click(logoutButton);

  expect(navigateMock).toHaveBeenCalledWith('/logout');
});

test('should change theme', async () => {
  const document = render(<GeneralSettingsView />);

  const themeDropdown = document.getByTestId(TestIds.THEME_DROPDOWN);
  await userEvent.selectOptions(themeDropdown, Theme.LIGHT);

  expect(setThemeMock).toHaveBeenCalledWith(Theme.LIGHT);
});

test('should change language', async () => {
  const changeLanguageMock = vi.spyOn(i18n, 'changeLanguage');

  const document = render(<GeneralSettingsView />);

  const languageDropdown = document.getByTestId(TestIds.LANGUAGE_DROPDOWN);
  await userEvent.selectOptions(languageDropdown, 'de');

  expect(changeLanguageMock).toHaveBeenCalledWith('de', expect.anything());
});
