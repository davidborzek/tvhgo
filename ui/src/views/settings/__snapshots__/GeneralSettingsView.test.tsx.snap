// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<DocumentFragment>
  <div
    class="_row_65046f"
  >
    <form
      class="_Form_27c918 _section_65046f"
    >
      <h2
        class="_headline_8da35a"
      >
        user
      </h2>
      <div
        class="_inputContainer_ae92e9"
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="username"
        >
          username
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9"
            name="username"
            placeholder="username"
            value="someUsername"
          />
        </div>
      </div>
      <div
        class="_inputContainer_ae92e9"
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="email"
        >
          email
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9"
            name="email"
            placeholder="email"
            value="some@email.com"
          />
        </div>
      </div>
      <div
        class="_inputContainer_ae92e9"
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="displayName"
        >
          display_name
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9"
            name="displayName"
            placeholder="display_name"
            value="Some User"
          />
        </div>
      </div>
      <div>
        <button
          class="_button_a7f9df _medium_a7f9df"
          data-testid="save-user-btn"
          type="submit"
        >
          save
        </button>
      </div>
    </form>
  </div>
  <div
    class="_row_65046f"
  >
    <div
      class="_section_65046f"
    >
      <h2
        class="_headline_8da35a"
      >
        user_interface
      </h2>
      <div
        class="_inputContainer_979188"
      >
        <label
          class="_inputLabel_979188"
        >
          appearance
        </label>
        <select
          class="_dropdown_979188"
          data-testid="theme-dropdown"
          style="width: 100%;"
        >
          <option
            value="dark"
          >
            dark
          </option>
          <option
            value="light"
          >
            light
          </option>
        </select>
      </div>
      <div
        class="_inputContainer_979188"
      >
        <label
          class="_inputLabel_979188"
        >
          language
        </label>
        <select
          class="_dropdown_979188"
          data-testid="language-dropdown"
          style="width: 100%;"
        >
          <option
            value="en"
          >
            english
          </option>
          <option
            value="de"
          >
            german
          </option>
          <option
            value="es"
          >
            spanish
          </option>
        </select>
      </div>
      <div
        class="_inputContainer_979188"
      >
        <label
          class="_inputLabel_979188"
        >
          time_format
        </label>
        <select
          class="_dropdown_979188"
          data-testid="time-format-dropdown"
          style="width: 100%;"
        >
          <option
            value="default"
          >
            default
          </option>
          <option
            value="en"
          >
            locales.en
          </option>
          <option
            value="af"
          >
            locales.af
          </option>
          <option
            value="ar-dz"
          >
            locales.ar-dz
          </option>
          <option
            value="ar-kw"
          >
            locales.ar-kw
          </option>
          <option
            value="ar-ly"
          >
            locales.ar-ly
          </option>
          <option
            value="ar-ma"
          >
            locales.ar-ma
          </option>
          <option
            value="ar-ps"
          >
            locales.ar-ps
          </option>
          <option
            value="ar-sa"
          >
            locales.ar-sa
          </option>
          <option
            value="ar-tn"
          >
            locales.ar-tn
          </option>
          <option
            value="ar"
          >
            locales.ar
          </option>
          <option
            value="az"
          >
            locales.az
          </option>
          <option
            value="be"
          >
            locales.be
          </option>
          <option
            value="bg"
          >
            locales.bg
          </option>
          <option
            value="bm"
          >
            locales.bm
          </option>
          <option
            value="bn-bd"
          >
            locales.bn-bd
          </option>
          <option
            value="bn"
          >
            locales.bn
          </option>
          <option
            value="bo"
          >
            locales.bo
          </option>
          <option
            value="br"
          >
            locales.br
          </option>
          <option
            value="bs"
          >
            locales.bs
          </option>
          <option
            value="ca"
          >
            locales.ca
          </option>
          <option
            value="cs"
          >
            locales.cs
          </option>
          <option
            value="cv"
          >
            locales.cv
          </option>
          <option
            value="cy"
          >
            locales.cy
          </option>
          <option
            value="da"
          >
            locales.da
          </option>
          <option
            value="de-at"
          >
            locales.de-at
          </option>
          <option
            value="de-ch"
          >
            locales.de-ch
          </option>
          <option
            value="de"
          >
            locales.de
          </option>
          <option
            value="dv"
          >
            locales.dv
          </option>
          <option
            value="el"
          >
            locales.el
          </option>
          <option
            value="en-au"
          >
            locales.en-au
          </option>
          <option
            value="en-ca"
          >
            locales.en-ca
          </option>
          <option
            value="en-gb"
          >
            locales.en-gb
          </option>
          <option
            value="en-ie"
          >
            locales.en-ie
          </option>
          <option
            value="en-il"
          >
            locales.en-il
          </option>
          <option
            value="en-in"
          >
            locales.en-in
          </option>
          <option
            value="en-nz"
          >
            locales.en-nz
          </option>
          <option
            value="en-sg"
          >
            locales.en-sg
          </option>
          <option
            value="eo"
          >
            locales.eo
          </option>
          <option
            value="es-do"
          >
            locales.es-do
          </option>
          <option
            value="es-mx"
          >
            locales.es-mx
          </option>
          <option
            value="es-us"
          >
            locales.es-us
          </option>
          <option
            value="es"
          >
            locales.es
          </option>
          <option
            value="et"
          >
            locales.et
          </option>
          <option
            value="eu"
          >
            locales.eu
          </option>
          <option
            value="fa"
          >
            locales.fa
          </option>
          <option
            value="fi"
          >
            locales.fi
          </option>
          <option
            value="fil"
          >
            locales.fil
          </option>
          <option
            value="fo"
          >
            locales.fo
          </option>
          <option
            value="fr-ca"
          >
            locales.fr-ca
          </option>
          <option
            value="fr-ch"
          >
            locales.fr-ch
          </option>
          <option
            value="fr"
          >
            locales.fr
          </option>
          <option
            value="fy"
          >
            locales.fy
          </option>
          <option
            value="ga"
          >
            locales.ga
          </option>
          <option
            value="gd"
          >
            locales.gd
          </option>
          <option
            value="gl"
          >
            locales.gl
          </option>
          <option
            value="gom-deva"
          >
            locales.gom-deva
          </option>
          <option
            value="gom-latn"
          >
            locales.gom-latn
          </option>
          <option
            value="gu"
          >
            locales.gu
          </option>
          <option
            value="he"
          >
            locales.he
          </option>
          <option
            value="hi"
          >
            locales.hi
          </option>
          <option
            value="hr"
          >
            locales.hr
          </option>
          <option
            value="hu"
          >
            locales.hu
          </option>
          <option
            value="hy-am"
          >
            locales.hy-am
          </option>
          <option
            value="id"
          >
            locales.id
          </option>
          <option
            value="is"
          >
            locales.is
          </option>
          <option
            value="it-ch"
          >
            locales.it-ch
          </option>
          <option
            value="it"
          >
            locales.it
          </option>
          <option
            value="ja"
          >
            locales.ja
          </option>
          <option
            value="jv"
          >
            locales.jv
          </option>
          <option
            value="ka"
          >
            locales.ka
          </option>
          <option
            value="kk"
          >
            locales.kk
          </option>
          <option
            value="km"
          >
            locales.km
          </option>
          <option
            value="kn"
          >
            locales.kn
          </option>
          <option
            value="ko"
          >
            locales.ko
          </option>
          <option
            value="ku-kmr"
          >
            locales.ku-kmr
          </option>
          <option
            value="ku"
          >
            locales.ku
          </option>
          <option
            value="ky"
          >
            locales.ky
          </option>
          <option
            value="lb"
          >
            locales.lb
          </option>
          <option
            value="lo"
          >
            locales.lo
          </option>
          <option
            value="lt"
          >
            locales.lt
          </option>
          <option
            value="lv"
          >
            locales.lv
          </option>
          <option
            value="me"
          >
            locales.me
          </option>
          <option
            value="mi"
          >
            locales.mi
          </option>
          <option
            value="mk"
          >
            locales.mk
          </option>
          <option
            value="ml"
          >
            locales.ml
          </option>
          <option
            value="mn"
          >
            locales.mn
          </option>
          <option
            value="mr"
          >
            locales.mr
          </option>
          <option
            value="ms-my"
          >
            locales.ms-my
          </option>
          <option
            value="ms"
          >
            locales.ms
          </option>
          <option
            value="mt"
          >
            locales.mt
          </option>
          <option
            value="my"
          >
            locales.my
          </option>
          <option
            value="nb"
          >
            locales.nb
          </option>
          <option
            value="ne"
          >
            locales.ne
          </option>
          <option
            value="nl-be"
          >
            locales.nl-be
          </option>
          <option
            value="nl"
          >
            locales.nl
          </option>
          <option
            value="nn"
          >
            locales.nn
          </option>
          <option
            value="oc-lnc"
          >
            locales.oc-lnc
          </option>
          <option
            value="pa-in"
          >
            locales.pa-in
          </option>
          <option
            value="pl"
          >
            locales.pl
          </option>
          <option
            value="pt-br"
          >
            locales.pt-br
          </option>
          <option
            value="pt"
          >
            locales.pt
          </option>
          <option
            value="ro"
          >
            locales.ro
          </option>
          <option
            value="ru"
          >
            locales.ru
          </option>
          <option
            value="sd"
          >
            locales.sd
          </option>
          <option
            value="se"
          >
            locales.se
          </option>
          <option
            value="si"
          >
            locales.si
          </option>
          <option
            value="sk"
          >
            locales.sk
          </option>
          <option
            value="sl"
          >
            locales.sl
          </option>
          <option
            value="sq"
          >
            locales.sq
          </option>
          <option
            value="sr-cyrl"
          >
            locales.sr-cyrl
          </option>
          <option
            value="sr"
          >
            locales.sr
          </option>
          <option
            value="ss"
          >
            locales.ss
          </option>
          <option
            value="sv"
          >
            locales.sv
          </option>
          <option
            value="sw"
          >
            locales.sw
          </option>
          <option
            value="ta"
          >
            locales.ta
          </option>
          <option
            value="te"
          >
            locales.te
          </option>
          <option
            value="tet"
          >
            locales.tet
          </option>
          <option
            value="tg"
          >
            locales.tg
          </option>
          <option
            value="th"
          >
            locales.th
          </option>
          <option
            value="tk"
          >
            locales.tk
          </option>
          <option
            value="tl-ph"
          >
            locales.tl-ph
          </option>
          <option
            value="tlh"
          >
            locales.tlh
          </option>
          <option
            value="tr"
          >
            locales.tr
          </option>
          <option
            value="tzl"
          >
            locales.tzl
          </option>
          <option
            value="tzm-latn"
          >
            locales.tzm-latn
          </option>
          <option
            value="tzm"
          >
            locales.tzm
          </option>
          <option
            value="ug-cn"
          >
            locales.ug-cn
          </option>
          <option
            value="uk"
          >
            locales.uk
          </option>
          <option
            value="ur"
          >
            locales.ur
          </option>
          <option
            value="uz-latn"
          >
            locales.uz-latn
          </option>
          <option
            value="uz"
          >
            locales.uz
          </option>
          <option
            value="vi"
          >
            locales.vi
          </option>
          <option
            value="x-pseudo"
          >
            locales.x-pseudo
          </option>
          <option
            value="yo"
          >
            locales.yo
          </option>
          <option
            value="zh-cn"
          >
            locales.zh-cn
          </option>
          <option
            value="zh-hk"
          >
            locales.zh-hk
          </option>
          <option
            value="zh-mo"
          >
            locales.zh-mo
          </option>
          <option
            value="zh-tw"
          >
            locales.zh-tw
          </option>
        </select>
      </div>
      <div>
        <button
          class="_button_a7f9df _medium_a7f9df _red_a7f9df"
          data-testid="logout-button"
        >
          logout
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render render when no user is present 1`] = `
<DocumentFragment>
  <div
    class="_Error_2b6812"
  >
    unexpected
  </div>
</DocumentFragment>
`;

exports[`should render render without logout button when authenticated by reverse proxy 1`] = `
<DocumentFragment>
  <div
    class="_row_65046f"
  >
    <form
      class="_Form_27c918 _section_65046f"
    >
      <h2
        class="_headline_8da35a"
      >
        user
      </h2>
      <div
        class="_inputContainer_ae92e9"
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="username"
        >
          username
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9"
            name="username"
            placeholder="username"
            value="someUsername"
          />
        </div>
      </div>
      <div
        class="_inputContainer_ae92e9"
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="email"
        >
          email
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9"
            name="email"
            placeholder="email"
            value="some@email.com"
          />
        </div>
      </div>
      <div
        class="_inputContainer_ae92e9"
        style="width: 100%;"
      >
        <label
          class="_inputLabel_ae92e9"
          for="displayName"
        >
          display_name
        </label>
        <div
          class="_innerContainer_ae92e9"
        >
          <input
            class="_input_ae92e9"
            name="displayName"
            placeholder="display_name"
            value="Some User"
          />
        </div>
      </div>
      <div>
        <button
          class="_button_a7f9df _medium_a7f9df"
          data-testid="save-user-btn"
          type="submit"
        >
          save
        </button>
      </div>
    </form>
  </div>
  <div
    class="_row_65046f"
  >
    <div
      class="_section_65046f"
    >
      <h2
        class="_headline_8da35a"
      >
        user_interface
      </h2>
      <div
        class="_inputContainer_979188"
      >
        <label
          class="_inputLabel_979188"
        >
          appearance
        </label>
        <select
          class="_dropdown_979188"
          data-testid="theme-dropdown"
          style="width: 100%;"
        >
          <option
            value="dark"
          >
            dark
          </option>
          <option
            value="light"
          >
            light
          </option>
        </select>
      </div>
      <div
        class="_inputContainer_979188"
      >
        <label
          class="_inputLabel_979188"
        >
          language
        </label>
        <select
          class="_dropdown_979188"
          data-testid="language-dropdown"
          style="width: 100%;"
        >
          <option
            value="en"
          >
            english
          </option>
          <option
            value="de"
          >
            german
          </option>
          <option
            value="es"
          >
            spanish
          </option>
        </select>
      </div>
      <div
        class="_inputContainer_979188"
      >
        <label
          class="_inputLabel_979188"
        >
          time_format
        </label>
        <select
          class="_dropdown_979188"
          data-testid="time-format-dropdown"
          style="width: 100%;"
        >
          <option
            value="default"
          >
            default
          </option>
          <option
            value="en"
          >
            locales.en
          </option>
          <option
            value="af"
          >
            locales.af
          </option>
          <option
            value="ar-dz"
          >
            locales.ar-dz
          </option>
          <option
            value="ar-kw"
          >
            locales.ar-kw
          </option>
          <option
            value="ar-ly"
          >
            locales.ar-ly
          </option>
          <option
            value="ar-ma"
          >
            locales.ar-ma
          </option>
          <option
            value="ar-ps"
          >
            locales.ar-ps
          </option>
          <option
            value="ar-sa"
          >
            locales.ar-sa
          </option>
          <option
            value="ar-tn"
          >
            locales.ar-tn
          </option>
          <option
            value="ar"
          >
            locales.ar
          </option>
          <option
            value="az"
          >
            locales.az
          </option>
          <option
            value="be"
          >
            locales.be
          </option>
          <option
            value="bg"
          >
            locales.bg
          </option>
          <option
            value="bm"
          >
            locales.bm
          </option>
          <option
            value="bn-bd"
          >
            locales.bn-bd
          </option>
          <option
            value="bn"
          >
            locales.bn
          </option>
          <option
            value="bo"
          >
            locales.bo
          </option>
          <option
            value="br"
          >
            locales.br
          </option>
          <option
            value="bs"
          >
            locales.bs
          </option>
          <option
            value="ca"
          >
            locales.ca
          </option>
          <option
            value="cs"
          >
            locales.cs
          </option>
          <option
            value="cv"
          >
            locales.cv
          </option>
          <option
            value="cy"
          >
            locales.cy
          </option>
          <option
            value="da"
          >
            locales.da
          </option>
          <option
            value="de-at"
          >
            locales.de-at
          </option>
          <option
            value="de-ch"
          >
            locales.de-ch
          </option>
          <option
            value="de"
          >
            locales.de
          </option>
          <option
            value="dv"
          >
            locales.dv
          </option>
          <option
            value="el"
          >
            locales.el
          </option>
          <option
            value="en-au"
          >
            locales.en-au
          </option>
          <option
            value="en-ca"
          >
            locales.en-ca
          </option>
          <option
            value="en-gb"
          >
            locales.en-gb
          </option>
          <option
            value="en-ie"
          >
            locales.en-ie
          </option>
          <option
            value="en-il"
          >
            locales.en-il
          </option>
          <option
            value="en-in"
          >
            locales.en-in
          </option>
          <option
            value="en-nz"
          >
            locales.en-nz
          </option>
          <option
            value="en-sg"
          >
            locales.en-sg
          </option>
          <option
            value="eo"
          >
            locales.eo
          </option>
          <option
            value="es-do"
          >
            locales.es-do
          </option>
          <option
            value="es-mx"
          >
            locales.es-mx
          </option>
          <option
            value="es-us"
          >
            locales.es-us
          </option>
          <option
            value="es"
          >
            locales.es
          </option>
          <option
            value="et"
          >
            locales.et
          </option>
          <option
            value="eu"
          >
            locales.eu
          </option>
          <option
            value="fa"
          >
            locales.fa
          </option>
          <option
            value="fi"
          >
            locales.fi
          </option>
          <option
            value="fil"
          >
            locales.fil
          </option>
          <option
            value="fo"
          >
            locales.fo
          </option>
          <option
            value="fr-ca"
          >
            locales.fr-ca
          </option>
          <option
            value="fr-ch"
          >
            locales.fr-ch
          </option>
          <option
            value="fr"
          >
            locales.fr
          </option>
          <option
            value="fy"
          >
            locales.fy
          </option>
          <option
            value="ga"
          >
            locales.ga
          </option>
          <option
            value="gd"
          >
            locales.gd
          </option>
          <option
            value="gl"
          >
            locales.gl
          </option>
          <option
            value="gom-deva"
          >
            locales.gom-deva
          </option>
          <option
            value="gom-latn"
          >
            locales.gom-latn
          </option>
          <option
            value="gu"
          >
            locales.gu
          </option>
          <option
            value="he"
          >
            locales.he
          </option>
          <option
            value="hi"
          >
            locales.hi
          </option>
          <option
            value="hr"
          >
            locales.hr
          </option>
          <option
            value="hu"
          >
            locales.hu
          </option>
          <option
            value="hy-am"
          >
            locales.hy-am
          </option>
          <option
            value="id"
          >
            locales.id
          </option>
          <option
            value="is"
          >
            locales.is
          </option>
          <option
            value="it-ch"
          >
            locales.it-ch
          </option>
          <option
            value="it"
          >
            locales.it
          </option>
          <option
            value="ja"
          >
            locales.ja
          </option>
          <option
            value="jv"
          >
            locales.jv
          </option>
          <option
            value="ka"
          >
            locales.ka
          </option>
          <option
            value="kk"
          >
            locales.kk
          </option>
          <option
            value="km"
          >
            locales.km
          </option>
          <option
            value="kn"
          >
            locales.kn
          </option>
          <option
            value="ko"
          >
            locales.ko
          </option>
          <option
            value="ku-kmr"
          >
            locales.ku-kmr
          </option>
          <option
            value="ku"
          >
            locales.ku
          </option>
          <option
            value="ky"
          >
            locales.ky
          </option>
          <option
            value="lb"
          >
            locales.lb
          </option>
          <option
            value="lo"
          >
            locales.lo
          </option>
          <option
            value="lt"
          >
            locales.lt
          </option>
          <option
            value="lv"
          >
            locales.lv
          </option>
          <option
            value="me"
          >
            locales.me
          </option>
          <option
            value="mi"
          >
            locales.mi
          </option>
          <option
            value="mk"
          >
            locales.mk
          </option>
          <option
            value="ml"
          >
            locales.ml
          </option>
          <option
            value="mn"
          >
            locales.mn
          </option>
          <option
            value="mr"
          >
            locales.mr
          </option>
          <option
            value="ms-my"
          >
            locales.ms-my
          </option>
          <option
            value="ms"
          >
            locales.ms
          </option>
          <option
            value="mt"
          >
            locales.mt
          </option>
          <option
            value="my"
          >
            locales.my
          </option>
          <option
            value="nb"
          >
            locales.nb
          </option>
          <option
            value="ne"
          >
            locales.ne
          </option>
          <option
            value="nl-be"
          >
            locales.nl-be
          </option>
          <option
            value="nl"
          >
            locales.nl
          </option>
          <option
            value="nn"
          >
            locales.nn
          </option>
          <option
            value="oc-lnc"
          >
            locales.oc-lnc
          </option>
          <option
            value="pa-in"
          >
            locales.pa-in
          </option>
          <option
            value="pl"
          >
            locales.pl
          </option>
          <option
            value="pt-br"
          >
            locales.pt-br
          </option>
          <option
            value="pt"
          >
            locales.pt
          </option>
          <option
            value="ro"
          >
            locales.ro
          </option>
          <option
            value="ru"
          >
            locales.ru
          </option>
          <option
            value="sd"
          >
            locales.sd
          </option>
          <option
            value="se"
          >
            locales.se
          </option>
          <option
            value="si"
          >
            locales.si
          </option>
          <option
            value="sk"
          >
            locales.sk
          </option>
          <option
            value="sl"
          >
            locales.sl
          </option>
          <option
            value="sq"
          >
            locales.sq
          </option>
          <option
            value="sr-cyrl"
          >
            locales.sr-cyrl
          </option>
          <option
            value="sr"
          >
            locales.sr
          </option>
          <option
            value="ss"
          >
            locales.ss
          </option>
          <option
            value="sv"
          >
            locales.sv
          </option>
          <option
            value="sw"
          >
            locales.sw
          </option>
          <option
            value="ta"
          >
            locales.ta
          </option>
          <option
            value="te"
          >
            locales.te
          </option>
          <option
            value="tet"
          >
            locales.tet
          </option>
          <option
            value="tg"
          >
            locales.tg
          </option>
          <option
            value="th"
          >
            locales.th
          </option>
          <option
            value="tk"
          >
            locales.tk
          </option>
          <option
            value="tl-ph"
          >
            locales.tl-ph
          </option>
          <option
            value="tlh"
          >
            locales.tlh
          </option>
          <option
            value="tr"
          >
            locales.tr
          </option>
          <option
            value="tzl"
          >
            locales.tzl
          </option>
          <option
            value="tzm-latn"
          >
            locales.tzm-latn
          </option>
          <option
            value="tzm"
          >
            locales.tzm
          </option>
          <option
            value="ug-cn"
          >
            locales.ug-cn
          </option>
          <option
            value="uk"
          >
            locales.uk
          </option>
          <option
            value="ur"
          >
            locales.ur
          </option>
          <option
            value="uz-latn"
          >
            locales.uz-latn
          </option>
          <option
            value="uz"
          >
            locales.uz
          </option>
          <option
            value="vi"
          >
            locales.vi
          </option>
          <option
            value="x-pseudo"
          >
            locales.x-pseudo
          </option>
          <option
            value="yo"
          >
            locales.yo
          </option>
          <option
            value="zh-cn"
          >
            locales.zh-cn
          </option>
          <option
            value="zh-hk"
          >
            locales.zh-hk
          </option>
          <option
            value="zh-mo"
          >
            locales.zh-mo
          </option>
          <option
            value="zh-tw"
          >
            locales.zh-tw
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
